### ISSUE TITLE
task: Train and test text2semantic under decoder only framework for ichigo v0.5

### ISSUE THREAD

**Issue by tikikun** [janhq/WhisperSpeech#3]
Since ichigo v0.5 will support additional language that will make the traditional t2s obsolete. This is a good chance to introduce a t2s framework that we have full control over.

# What needed to be done:
-  [ ] Train t2s on a decoder model to test feasibility (we already have the data) under ichigo v0.4
-  [ ] Train t2s on decoder model on new quantizer after janhq/WhisperSpeech#2 is done

**Comment by bachvudinh** [comment]
## Experiments

| Run ID | Date       | Model Config | Dataset | Learning Rate | Batch Size | Steps | Loss | Hardware | MMLU | MMLU pro | Notes |
|--------|------------|--------------|---------|---------------|------------|-------|------|----------|------|-----------|--------|
| exp1-pretrain | 2024-11-23 | [Lora-256-512](https://github.com/janhq/research/blob/main/assets/ichigo-32b/exp1/config/32B_lora_pretrain.yaml) | [Pretrain v0.1](https://huggingface.co/datasets/homebrewltd/raw-speech-whispervq-v1) | 1.5e-4 | 384  | 6302 | [1.9](https://github.com/janhq/research/blob/main/assets/ichigo-32b/exp1/loss/qwen2.5-32b-ichigo-sft.png) | ~ 100 hours on 8xA6000 | - | - | old dataset |

**Comment by hahuyhoang411** [comment]
I move the table to the top for better visualization cc @bachvudinh

**Comment by PodsAreAllYouNeed** [comment]
This task is a hybrid between Text-to-speech and speech-to-speech translation. It is quite hard because there is a one-to-many mapping between input text, and possible output token combinations. 

Here are two papers that are using the same AR setting, but for slightly different tasks. I think it can be adapted. 

AudioPALM: https://arxiv.org/pdf/2306.12925
Valle-E: https://arxiv.org/pdf/2301.02111

Specifically, I think we can use Valle-E's idea of using a phoneme conversion layer before sending the text into the AR model, this might bridge the gap to the semantic embeddings abit, making the AR model's job easier. We also need to somehow provide some auxiliary information about the expected acoustic ground-truth that we are using, otherwise, if we provide text-only to the AR model, there are too many possible correct answers, so across multiple samples the loss may conflict. 

However, I think it will be hard to make this work. The AR model needs a better constraint. 

### My proposal

In the WhisperSpeech framework, the text-to-semantic model is the inverse of the whisper decoder. We need to involve the whisper decoder in the training. 

1) Keep the same AR model structure
2) However, instead of trying to get the model to predict the whipserVQ codes, send _continuous embeddings_ into the frozen whisper decoder. What we are trying to do is get the AR decoder model to trick the whisper decoder into thinking it is seeing output from the whisper encoder. 
3) Compute the loss of the whisper decoder output to the original text. 

You will meet a practical challenge, which is that while training this AR decoder model, its acting like its a NAR encoder model to the Whisper Decoder. There might be a smart way to solve this, but I can't think of one at the moment, or you can just use a NAR model. 

### Another (Simpler) Idea

If we really want an AR model trained using next token prediction, we must use WhisperVQ tokens in the current format and we don't want to add auxiliary information, we can try a simple intervention of grouping identical WhisperVQ tokens together. This way, the model is not penalized for getting the output length wrong.

i.e this original example:
<|sound_start|><|sound_0209|><|sound_0134|><|sound_0134|><|sound_0134|><|sound_0241|><|sound_0222|><|sound_0239|><|sound_0329|><|sound_0197|><|sound_0115|><|sound_0409|><|sound_0196|><|sound_0196|><|sound_0196|><|sound_0196|><|sound_0196|><|sound_0235|><|sound_0487|><|sound_0487|><|sound_0459|><|sound_0459|><|sound_0405|><|sound_end|> 

get mapped to this:
<|sound_start|><|sound_0209|><|sound_0134|><|sound_0241|><|sound_0222|><|sound_0239|><|sound_0197|><|sound_0115|><|sound_0409|><|sound_0196|><|sound_0235|><|sound_0487|><|sound_0459|><|sound_0405|><|sound_end|> 

This way the order of the token output matters, but the number of consecutively repeated tokens do not matter.
We can worry about upsampling the number of tokens as a separate problem. It might not matter to the decoder because the whole token sequence gets cross-attention anyway. The repeated tokens might not be adding that much information. During Fine-tuning we can apply a similar filtering to the WhisperVQ token stream, to see if the performance changes. Actually, if repeated tokens don't impact performance, then actually it makes inference even faster.

**Comment by bachvudinh** [comment]
Updated from Research sync 2024-12-4:
- The current model have repetitive problem that generate endless sound semantic token and don't know how to generate <|sound_end|> token to stop.
- The latest on-going  runs that we scale the data to 3.7M samples, change the model to llama3.1 1B base version and doing 3 epoch. The loss metric stabilizes between 1.4-1.5 after the second epoch and shows no further improvement during the third epoch
- @PodsAreAllYouNeed suggested that we should deduplicate sound token when making training data. The repeated tokens might not be adding that much information and its partition to how fast a work is pronounce.

**Comment by bachvudinh** [comment]
cc @PodsAreAllYouNeed @tikikun 
I do some testing with whisperVQ to see if the number of repeated semantic token proportion to how fast or slow a word are pronounce:
- Hello World Fast:
  Semantic token: 
  ```
<|sound_0314|><|sound_0336|><|sound_0500|><|sound_0207|><|sound_0207|><|sound_0153|><|sound_0013|><|sound_0208|><|sound_0013|><|sound_0013|><|sound_0137|><|sound_0228|><|sound_0137|><|sound_0228|><|sound_0228|><|sound_0228|><|sound_0322|><|sound_0378|><|sound_0407|><|sound_0407|><|sound_0407|><|sound_0378|><|sound_0200|><|sound_0200|><|sound_0407|><|sound_0378|><|sound_0200|><|sound_0378|><|sound_0200|><|sound_0200|><|sound_0407|><|sound_0200|><|sound_0407|><|sound_0078|><|sound_0407|><|sound_0378|><|sound_0407|><|sound_0378|><|sound_0200|><|sound_0407|><|sound_0200|><|sound_0200|><|sound_0192|><|sound_0407|><|sound_0407|><|sound_0407|><|sound_0192|><|sound_0192|><|sound_0397|><|sound_0397|><|sound_0397|><|sound_0479|><|sound_0341|><|sound_0479|><|sound_0245|><|sound_0245|><|sound_0245|><|sound_0245|><|sound_0245|><|sound_0035|><|sound_0056|><|sound_0245|><|sound_0400|><|sound_0376|><|sound_0446|><|sound_0378|><|sound_0400|><|sound_0192|><|sound_0192|><|sound_0400|><|sound_0200|><|sound_0200|><|sound_0213|><|sound_0207|><|sound_0200|><|sound_0207|><|sound_0200|><|sound_0207|><|sound_0207|><|sound_0207|><|sound_0200|><|sound_0508|><|sound_0508|><|sound_0508|><|sound_0508|><|sound_0207|><|sound_0336|><|sound_0348|><|sound_0336|><|sound_0508|><|sound_0336|><|sound_0508|><|sound_0200|>
  ```

**Comment by bachvudinh** [comment]
what I tried to do:
- Deduplicate the sound semantic token to get the shorter sound token sequence --> Efficient Training. I experimented with Llama 3.1 1B for 3 epoch with [deduplicated data](https://huggingface.co/datasets/jan-hq/instruction-convert-audio-whispervq-llama3.2-dedup)
 for example:
 Input:
 ```
 <|sound_start|><|sound_0207|><|sound_0148|><|sound_0000|><|sound_0249|><|sound_0249|><|sound_0177|><|sound_0177|><|sound_0177|><|sound_0177|><|sound_0177|><|sound_0302|><|sound_0302|><|sound_0426|><|sound_0426|><|sound_0081|><|sound_0090|><|sound_0491|><|sound_0327|><|sound_0327|><|sound_0153|><|sound_0061|><|sound_0061|><|sound_0196|><|sound_0178|><|sound_0129|><|sound_0129|><|sound_0025|><|sound_0070|><|sound_0114|><|sound_0159|><|sound_0159|><|sound_0159|><|sound_0010|><|sound_0316|><|sound_0270|><|sound_0510|><|sound_0409|><|sound_0339|><|sound_0378|><|sound_0339|><|sound_0378|><|sound_0407|><|sound_0407|><|sound_0404|><|sound_0404|><|sound_0235|><|sound_0322|><|sound_0322|><|sound_0076|><|sound_0245|><|sound_0226|><|sound_0011|><|sound_0407|><|sound_0407|><|sound_0007|><|sound_0007|><|sound_0007|><|sound_0187|><|sound_0316|><|sound_0300|><|sound_0316|><|sound_0316|><|sound_0109|><|sound_0008|><|sound_0300|><|sound_0179|><|sound_0179|><|sound_0179|><|sound_0122|><|sound_0122|><|sound_0158|><|sound_0158|><|sound_0245|><|sound_0080|><|sound_0080|><|sound_0461|><|sound_0461|><|sound_0005|><|sound_0368|><|sound_0368|><|sound_0487|><|sound_0206|><|sound_0278|><|sound_0278|><|sound_0002|><|sound_0002|><|sound_0102|><|sound_0329|><|sound_0444|><|sound_0114|><|sound_0114|><|sound_0325|><|sound_0202|><|sound_0202|><|sound_0202|><|sound_0363|><|sound_0319|><|sound_0161|><|sound_0147|><|sound_0397|><|sound_0325|><|sound_0081|><|sound_0081|><|sound_0319|><|sound_0161|><|sound_0331|><|sound_0187|><|sound_0350|><|sound_0210|><|sound_0048|><|sound_0468|><|sound_0468|><|sound_0125|><|sound_0271|><|sound_0271|><|sound_0157|><|sound_0344|><|sound_0007|><|sound_0300|><|sound_0300|><|sound_0300|><|sound_0300|><|sound_0300|><|sound_0076|><|sound_0226|><|sound_0498|><|sound_0498|><|sound_0330|><|sound_0468|><|sound_0468|><|sound_0468|><|sound_0434|><|sound_0182|><|sound_end|>
 ```
 output:
 ```
 <|sound_start|><|sound_0207|><|sound_0148|><|sound_0000|><|sound_0249|><|sound_0177|><|sound_0302|><|sound_0426|><|sound_0081|><|sound_0090|><|sound_0491|><|sound_0327|><|sound_0153|><|sound_0061|><|sound_0196|><|sound_0178|><|sound_0129|><|sound_0025|><|sound_0070|><|sound_0114|><|sound_0159|><|sound_0010|><|sound_0316|><|sound_0270|><|sound_0510|><|sound_0409|><|sound_0339|><|sound_0378|><|sound_0339|><|sound_0378|><|sound_0407|><|sound_0404|><|sound_0235|><|sound_0322|><|sound_0076|><|sound_0245|><|sound_0226|><|sound_0011|><|sound_0407|><|sound_0007|><|sound_0187|><|sound_0316|><|sound_0300|><|sound_0316|><|sound_0109|><|sound_0008|><|sound_0300|><|sound_0179|><|sound_0122|><|sound_0158|><|sound_0245|><|sound_0080|><|sound_0461|><|sound_0005|><|sound_0368|><|sound_0487|><|sound_0206|><|sound_0278|><|sound_0002|><|sound_0102|><|sound_0329|><|sound_0444|><|sound_0114|><|sound_0325|><|sound_0202|><|sound_0363|><|sound_0319|><|sound_0161|><|sound_0147|><|sound_0397|><|sound_0325|><|sound_0081|><|sound_0319|><|sound_0161|><|sound_0331|><|sound_0187|><|sound_0350|><|sound_0210|><|sound_0048|><|sound_0468|><|sound_0125|><|sound_0271|><|sound_0157|><|sound_0344|><|sound_0007|><|sound_0300|><|sound_0076|><|sound_0226|><|sound_0498|><|sound_0330|><|sound_0468|><|sound_0434|><|sound_0182|><|sound_end|>
 ```
 Result:
 - The loss did not go lower than 1.7 after 2 epoch (~ 1.73-1.8) ---> Model is unable to identify and learn the data pattern, possibly due to information loss caused by deduplication.

**Comment by tikikun** [comment]
# Idea: Add duration tokens

Observations:
- Loss did not converge as well after we de-dup, something must be lost
- Repetition is the only difference, how many time a sound token is repeated
- Continuous embedding is a by product of how Whisper Encoder process (each embedding is slightly different from other one even though they are almost the same)

Theories:
- We need to be able to "add back the "repetition" information
- After we make sure the information is kept the same, the loss behaviour should be relatively the same.

Implementation:
- We create "<|dur_xxx|> (with x is an integer) to represent the repetition inside each repeated sound tokens group

![Image](https://github.com/user-attachments/assets/182da3c3-7d27-4247-a48e-dc143f43ef75)

**Comment by PodsAreAllYouNeed** [comment]
Related to the idea of the token-level duration token, we could potentially have a "global duration token" added as a context token either before or after the provided text input. This "global duration token" gives information to the t2s model about the length of semantic tokens it needs to generate. After training, this global duration token is also used to control the length of generation of the text, which controls the speaking speed. This is inspired by the "number of frames" mechanic found in the F5-TTS code generation [https://github.com/SWivid/F5-TTS/blob/8898d05e374bcb8d3fc0b1286037e95df61f491f/src/f5_tts/infer/utils_infer.py#L449C1-L452C96](https://github.com/SWivid/F5-TTS/blob/8898d05e374bcb8d3fc0b1286037e95df61f491f/src/f5_tts/infer/utils_infer.py#L449C1-L452C96.)

If TTS models need some global duration information in order to do the generation, then our text2semantic should also use the same kind of global information. we just need to encode it a little differently.

**Comment by bachvudinh** [comment]
Some result from the training runs:
 - The loss going lower than normal deduplication training ---> "repetition" information
 - The loss does not converge better than training using normal sound token ---> this maybe due to new added "duration token"  
 ![image](https://github.com/user-attachments/assets/7011baf3-4695-4abb-bb4e-bc7f67b66b4c)
